edge/pkg/edgehub/upgrade/upgrade.go

```diff
@@ -128,14 +135,27 @@ func (*keadmUpgrade) Upgrade(upgradeReq *commontypes.NodeUpgradeJobRequest) erro
 	files := map[string]string{
 		filepath.Join(util.KubeEdgeUsrBinPath, util.KeadmBinaryName): filepath.Join(util.KubeEdgeUsrBinPath, util.KeadmBinaryName),
 	}
-	err = container.CopyResources(image, files)
-	if err != nil {
-		return fmt.Errorf("failed to cp file from image to host: %v", err)
+
+	//  the path name for the executable that started the current edgecore process
+	binaryPath, _ := os.Executable()
+	// no need to copy keadm in e2e test, we should use the new compiled keadm
+	if !upgrade.fake {
+		err = container.CopyResources(image, files)
+		if err != nil {
+			return fmt.Errorf("failed to cp file from image to host: %v", err)
+		}
 	}
 
 	klog.Infof("Begin to run upgrade command")
-	upgradeCmd := fmt.Sprintf("keadm upgrade --upgradeID %s --historyID %s --fromVersion %s --toVersion %s --config %s --image %s > /tmp/keadm.log 2>&1",
-		upgradeReq.UpgradeID, upgradeReq.HistoryID, version.Get(), upgradeReq.Version, opts.ConfigFile, image)
+	upgradeCmd := fmt.Sprintf("keadm upgrade --upgradeID %s --historyID %s --fromVersion %s --toVersion %s --config %s --image %s --binaryPath %s > /tmp/keadm.log 2>&1",
+		upgradeReq.UpgradeID, upgradeReq.HistoryID, version.Get(), upgradeReq.Version, opts.ConfigFile, image, binaryPath)
+
+	// only used in test
+	if upgrade.fake {
+		// add a `--fake=true` flag to indicate that we'll not do actual Upgrade operations
```

Comment:
why not use a dry-run flag, I think it more common.

Reply:
ok, will fix this